### PR TITLE
Optimize stomp encoder

### PR DESCRIPTION
**Pull request by timandy**
Modification:
- Use unpooled composite contentBuf and NULL to reduce memory alloc and copy.
- In stomp over websocket, Two message object will send two TextWebSocketFrame. To reduce bandwidth usage, we should merge these two messages.

### CODE REVIEW

File: codec-stomp/src/main/java/io/netty/handler/codec/stomp/StompSubframeEncoder.java [hunk 1]
```diff
@@ -28,16 +29,18 @@
...
     @Override
     protected void encode(ChannelHandlerContext ctx, StompSubframe msg, List<Object> out) throws Exception {
         if (msg instanceof StompFrame) {
             StompFrame frame = (StompFrame) msg;
             ByteBuf frameBuf = encodeFrame(frame, ctx);
             if (frame.content().isReadable()) {
-                out.add(frameBuf);
                 ByteBuf contentBuf = encodeContent(frame, ctx);
-                out.add(contentBuf);
+                ByteBuf buf = Unpooled.wrappedBuffer(frameBuf, contentBuf);
```
Comment (normanmaurer) [inline codec-stomp/src/main/java/io/netty/handler/codec/stomp/StompSubframeEncoder.java:42]: Why you think this a good change ?

Comment (timandy) [reply codec-stomp/src/main/java/io/netty/handler/codec/stomp/StompSubframeEncoder.java:42]: @normanmaurer In my project stomp over websocket. If `StompSubframEncodeer` return two msg, I have to aggregate the two msg into one `WebSocketTextFrame` in my `StompOverWebsocketAggregator`. If it return only one msg, I just wrap the msg with `WebSocketTextFrame`.

ps: In my project write DefaultStompFrame only.


My StompOverWebsocketAggregator
```
    @Override
    protected void encode(ChannelHandlerContext ctx, ByteBuf msg, List<Object> out) {
        if (msg == null)
            return;
        out.add(new TextWebSocketFrame(msg.retain()));
    }
```

**Comment by timandy** [comment]
> I am not sure this PR really improves anything to be honest.

@normanmaurer Compsite ByteBuf is faster than alloc and copy.

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by normanmaurer** [comment]
@timandy there are test failures

**Comment by timandy** [comment]
@netty-bot test this please

**Comment by timandy** [comment]
@normanmaurer  Sorry for my late. I fixed the testcase.  PTAL.

**Comment by chrisvest** [comment]
@netty-bot test this please

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by amizurov** [comment]
@normanmaurer I'll take a look today

**Comment by amizurov** [comment]
@normanmaurer Unfortunately we have performance regression with this PR, i'll spend more time to investigate what we can to do.
Current `StompSubframeEncoder` and `PooledByteBufAllocator`.
```
Benchmark                              (contentLength)   Mode  Cnt        Score        Error  Units
StompEncoderBenchmark.writeStompFrame                0  thrpt   10  2704023.816 ±  64106.901  ops/s
StompEncoderBenchmark.writeStompFrame               10  thrpt   10  2597182.372 ±  13783.263  ops/s
StompEncoderBenchmark.writeStompFrame              100  thrpt   10  2680736.401 ±  20726.501  ops/s
StompEncoderBenchmark.writeStompFrame              500  thrpt   10  3010725.566 ±  12284.409  ops/s
StompEncoderBenchmark.writeStompFrame             1000  thrpt   10  2991846.248 ±  15521.308  ops/s
StompEncoderBenchmark.writeStompFrame             2500  thrpt   10  2688040.108 ±  17355.140  ops/s
StompEncoderBenchmark.writeStompFrame             5000  thrpt   10  2801412.646 ± 163001.899  ops/s
StompEncoderBenchmark.writeStompFrame            10000  thrpt   10  2924669.216 ±  20573.418  ops/s
StompEncoderBenchmark.writeStompFrame            50000  thrpt   10  3076656.711 ±  17770.302  ops/s
StompEncoderBenchmark.writeStompFrame           100000  thrpt   10  2721926.655 ± 119518.510  ops/s
```

Proposed `StompSubframeEncoder` and `PooledByteBufAllocator`.
```
Benchmark                              (contentLength)   Mode  Cnt        Score        Error  Units
StompEncoderBenchmark.writeStompFrame                0  thrpt   10  2539086.748 ± 236109.776  ops/s
StompEncoderBenchmark.writeStompFrame               10  thrpt   10  1598682.142 ±  34474.200  ops/s
StompEncoderBenchmark.writeStompFrame              100  thrpt   10  1657313.955 ±  37222.967  ops/s
StompEncoderBenchmark.writeStompFrame              500  thrpt   10  1615968.008 ±   7541.027  ops/s
StompEncoderBenchmark.writeStompFrame             1000  thrpt   10  1618509.670 ±   3929.177  ops/s
StompEncoderBenchmark.writeStompFrame             2500  thrpt   10  1583537.859 ±   5722.270  ops/s
StompEncoderBenchmark.writeStompFrame             5000  thrpt   10  1590876.384 ±  12699.172  ops/s
StompEncoderBenchmark.writeStompFrame            10000  thrpt   10  1556409.938 ±   2672.576  ops/s
StompEncoderBenchmark.writeStompFrame            50000  thrpt   10  1631375.845 ±   3847.993  ops/s
StompEncoderBenchmark.writeStompFrame           100000  thrpt   10  1687448.349 ±   3873.697  ops/s
```

**Comment by amizurov** [comment]
PTAL https://github.com/netty/netty/pull/10778

**Comment by franz1981** [comment]
hi @amizurov but I cannot find StompEncoderBenchmark, where it is?

**Comment by amizurov** [comment]
> hi @amizurov but I cannot find StompEncoderBenchmark, where it is?
> 
> it's https://github.com/netty/netty/pull/10778/files#diff-d5944d95d897fcd0e93a249cb2c815a555ce88bcd86684f7823109a1e60f9e97?

Hi, yep. I'll be glad if you look.

**Comment by franz1981** [comment]
@amizurov A bit OT, but I've recently improved single byte searchs on https://github.com/netty/netty/pull/10737: I see that Stomp decoding is based on a appending NUL at the end of a frame (https://github.com/netty/netty/blob/e5951d46fc89db507ba7d2968d2ede26378f0b04/codec-stomp/src/main/java/io/netty/handler/codec/stomp/StompSubframeDecoder.java#L148). Maybe you're interested to check if my changes will positively (or the opposite) affect your use case with Stomp ;)

> Hi, yep. I'll be glad if you look.
Yep!

**Comment by amizurov** [comment]
> @amizurov A bit OT, but I've recently improved single byte searchs on #10737: I see that Stomp decoding is based on a appending NUL at the end of a frame (
> 
> https://github.com/netty/netty/blob/e5951d46fc89db507ba7d2968d2ede26378f0b04/codec-stomp/src/main/java/io/netty/handler/codec/stomp/StompSubframeDecoder.java#L148
> 
> ). Maybe you're interested to check if my changes will positively (or the opposite) affect your use case with Stomp ;)
> > Hi, yep. I'll be glad if you look.
> > Yep!

We can add benchmark for decoder. If `content-length` header is not present we trying to found NULL octet for calculating a payload size.

**Comment by timandy** [comment]
Close this PR. Cause #10778  solved the problem.

**Pull request closed without merging**